### PR TITLE
Print to PDF

### PR DESCRIPTION
**Pull request by signedav**
When a project has print layouts, they are available on QField as well.

It's tested on a mobile phone, that's why the small screen:
print2

On a large amount of layouts, it would be still comfortable to handle.

printing

So the basic works. If the print layouts looks the same on device than on the computer needs to be validated still.

Nice things would be: 
- Busy indicator (because old devices are slow) - the implementation in this PR is not yet working.
- Temporary file handling (that files are deleted afterwards)

### CODE REVIEW

**Comment by signedav** [comment]
@m-kuhn should the print()-function be in the printlayoutmodel or is the qgismobileapp the correct place?

File: src/printlayoutlistmodel.cpp [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+  reloadModel();
+}
+
+void PrintLayoutListModel::reloadModel()
+{
+  beginResetModel();
+  mPrintLayouts.clear();
+
+  for ( const auto &layout : mProject->layoutManager()->printLayouts() )
```
Comment (m-kuhn) [inline src/printlayoutlistmodel.cpp]: Take a const copy of the list to avoid a detach

File: src/printlayoutlistmodel.cpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  }
+  endResetModel();
+
+}
+
+int PrintLayoutListModel::rowCount( const QModelIndex& parent ) const
+{
+  Q_UNUSED( parent )
+  int test = mPrintLayouts.size();
+  return test;
```
Comment (m-kuhn) [inline src/printlayoutlistmodel.cpp]: Why the test? :)

File: src/printlayoutlistmodel.cpp [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+QVariant PrintLayoutListModel::data( const QModelIndex& index, int role ) const
+{
+  if ( index.row() >= mPrintLayouts.size() )
```
Comment (m-kuhn) [inline src/printlayoutlistmodel.cpp]: or < 0?

File: src/printlayoutlistmodel.h [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+#include <qgslayoutpagecollection.h>
+
+class QStringList;
+class QgsProject;
+
+class PrintLayoutListModel : public QAbstractListModel
+{
+    Q_OBJECT
+
+    Q_PROPERTY( QgsProject* project READ project WRITE setProject )
```
Comment (m-kuhn) [inline src/printlayoutlistmodel.h]: If in doubt, also do a NOTIFY

File: src/printlayoutlistmodel.h [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+    enum Roles
+    {
+      TitleRow,
```
Comment (m-kuhn) [inline src/printlayoutlistmodel.h]: Can you add some (very short) docstrings

File: src/qgismobileapp.cpp [hunk 6]
```diff
@@ -268,6 +278,36 @@ void QgisMobileapp::loadProjectFile( const QString& path )
   emit loadProjectEnded();
 }
 
+void QgisMobileapp::print( int layoutIndex )
+{
+  const QList<QgsPrintLayout *> projectLayouts( mProject->layoutManager()->printLayouts() );
+
+  const auto &layoutToPrint = projectLayouts.at( layoutIndex );
```
Comment (m-kuhn) [inline src/qgismobileapp.cpp]: This will crash if the index is out of bounds. Not sure if you caught that somewhere else

File: src/qgismobileapp.cpp [hunk 7]
```diff
@@ -268,6 +278,36 @@ void QgisMobileapp::loadProjectFile( const QString& path )
...
+
+  emit printingStarted( layoutToPrint->name() );
+
+  if ( layoutToPrint->pageCollection()->pageCount() == 0 )
+    return;
+
+  QPrinter printer;
+  printer.setOutputFormat(QPrinter::PdfFormat);
+  printer.setPaperSize(QPrinter::A4);
```
Comment (m-kuhn) [inline src/qgismobileapp.cpp]: Is this required? I think that's defined by the layout itself

File: src/qgismobileapp.cpp [hunk 8]
```diff
@@ -268,6 +278,36 @@ void QgisMobileapp::loadProjectFile( const QString& path )
...
+  printer.setOutputFileName( mProject->homePath() + "/" + layoutToPrint->name() + ".pdf" );
```
Comment (m-kuhn) [inline src/qgismobileapp.cpp]: QStringLiteral for the ".pdf", '/' for the "/"

**Comment by qfield-fairy** [comment]
Uploaded test apks for [armv7](https://opengis.ch/download/qfield/ci-builds/qfield-dev-324-cafedc20caac2274ac7d7fad07818d10f31a9249-armv7.apk) and [x86](https://opengis.ch/download/qfield/ci-builds/qfield-dev-324-cafedc20caac2274ac7d7fad07818d10f31a9249-x86.apk)

**Comment by m-kuhn** [comment]
Good job :tada: Love it!!

**Comment by m-kuhn** [comment]
> Nice things would be:
> 
>    Busy indicator (because old devices are slow) - the implementation in this PR is not yet working.

Right, that would be nice. I don't know the API for this.

>    Temporary file handling (that files are deleted afterwards)

I think you convinced me that it is even better without this

**Comment by signedav** [comment]
Regarding the Busy Indicator I have some trouble.

On the current implementation, where the `QgisMobileapp::print` emits signals (`printingStarted` on start and printingEnded` on end), there is nothing happening on the BusyIndicator `busyMessageIndicator` and it's rectangle `busyMessage`. Why? Because there seems to be no visual updates while something is happening in `QgisMobileapp::print`, like the Thread would be blocked (but the signals are received).
I belief that this is on loading Projects as well, so the old code would work neither. 
Do you know a sollution for this @m-kuhn or @3nids ?

I read about `WorkerScripts` but there we will need to call the `QgisMobileapp` from a `.js` File, and I'm not sure about all the consequences there...

PS. The BusyIndicator is mainly important on slow devices. While my Galaxy S6 Mobile Phone is quite fast on our Active Tap it's taking around 12 seconds to print a layout...

File: src/qml/qgismobileapp.qml [hunk 9]
```diff
@@ -809,8 +808,8 @@ ApplicationWindow {
   Controls.BusyIndicator {
     id: busyIndicator
     anchors { right: alertIcon.left; top: parent.top }
-    width: 36 * dp
-    height: 36 * dp
+    width: 96 * dp
```
Comment (m-kuhn) [inline src/qml/qgismobileapp.qml]: Is this related to this change?
I think it's unfortunately not that simple, it looks good on low-res devices and bad on high-res devices, so I think we need to do something else than just increase the factor. Can we just keep it out for now?

Comment (signedav) [reply src/qml/qgismobileapp.qml]: Yes. Actually it has been more for size testing. We can remove it.

File: src/printlayoutlistmodel.cpp [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+QgsProject* PrintLayoutListModel::project() const
+{
+  return mProject;
+}
+
+void PrintLayoutListModel::setProject( QgsProject* project )
+{
+  mProject = project;
```
Comment (m-kuhn) [inline src/printlayoutlistmodel.cpp]: Better add a guard here

```if ( mProject == project ) 
return;
?```

File: src/printlayoutlistmodel.cpp [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+    mPrintLayouts.append( PrintLayout( layout->name() ) );
+  }
+  endResetModel();
+
+}
+
+int PrintLayoutListModel::rowCount( const QModelIndex& parent ) const
+{
+  Q_UNUSED( parent )
+  return mPrintLayouts.size();
```
Comment (m-kuhn) [inline src/printlayoutlistmodel.cpp]:
```
if ( !parent.isValid() )
   return layouts.size()
else
  return 0;
```

File: src/printlayoutlistmodel.h [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+    Q_PROPERTY( QgsProject* project READ project WRITE setProject NOTIFY projectChanged )
+
+     /*!
+      * Roles to get the data of the model.
+      * To get the printLayout attributes TitleRow is used.
+      * To get position in the list of PrintLayouts IndexRow is used.
+     */
+    enum Roles
+    {
+      TitleRow,
```
Comment (m-kuhn) [inline src/printlayoutlistmodel.h]: You can add docstrings to individual items

```
  TitleRow, //!< The title of a printLayout
```

File: src/printlayoutlistmodel.h [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+      IndexRow
+    };
+
+    /*!
+     * Struct of printLayout attributes (at the moment only for the title (name) used)
+    */
+    //! at the moment only contains the title string - probably in future more attributes needed
```
Comment (m-kuhn) [inline src/printlayoutlistmodel.h]: I think you can remove this line, that should be obvious ;)

**Comment by m-kuhn** [comment]
> BusyIndicatorbusyMessageIndicatorand it's rectanglebusyMessage. Why? Because there seems to be no visual updates while something is happening inQgisMobileapp::print`, like the Thread would be blocked (but the signals are received).

I guess the printing code is not safe to be sent to a background thread. I'd just accept the deadlock for now. Without digging a bit deeper I would be worried to send it to a background thread.

**Comment by signedav** [comment]
Hm... Okay, I understand. 
Are you aware that the BusyIndicator on opening files does not work neither or is it working and I missed something?

**Comment by m-kuhn** [comment]
Yes, makes sense, that it's the same issue there.
Can you have a look at the last comments, I'd like to merge this soon!

File: src/printlayoutlistmodel.cpp (hunk 11 above)
Comment (signedav) [reply src/printlayoutlistmodel.cpp]: I've seen it in the QAbstractItemModel, that it should return 0. But I don't understand why...

File: src/printlayoutlistmodel.cpp (hunk 10 above)
Comment (signedav) [reply src/printlayoutlistmodel.cpp]: Because it's used for refreshing model as well, I would prefer:
```
  if( mProject != project )
    mProject = project;

  reloadModel();
}
```
What do you think?

Comment (m-kuhn) [reply src/printlayoutlistmodel.cpp]: Why not call `reloadModel()` explicitly in such cases?

File: src/printlayoutlistmodel.cpp (hunk 11 above)
Comment (m-kuhn) [reply src/printlayoutlistmodel.cpp]: models are capable of providing tree-structures. So you can call this for an item at row 1 and it will give you the number of nodes http://doc.qt.io/qt-5/qtwidgets-itemviews-simpletreemodel-example.html#design-and-concepts

Comment (signedav) [reply src/printlayoutlistmodel.cpp]: Okay, I understand. Because it request of a table the amount of nonexistent children of an entry. Thanks for clarification.

File: src/printlayoutlistmodel.cpp (hunk 10 above)
Comment (signedav) [reply src/printlayoutlistmodel.cpp]: I thought because it needs to reload after loading project anyway. But I can do that and then there are two calls onLoadProjectEnded. No problem.

**Comment by qfield-fairy** [comment]
Uploaded test apks for [armv7](https://opengis.ch/download/qfield/ci-builds/qfield-dev-324-b4a268c42d10930fba7c9a77864332a882d29e8e-armv7.apk) and [x86](https://opengis.ch/download/qfield/ci-builds/qfield-dev-324-b4a268c42d10930fba7c9a77864332a882d29e8e-x86.apk)
